### PR TITLE
Aggregation Pushdown for Druid connector

### PR DESCRIPTION
**Pull request by zhenxiao**
```
== RELEASE NOTES ==

Druid Changes
* Aggregation Pushdown for Druid connector
```

### CODE REVIEW

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+{
+    private static final Map<String, String> PRESTO_TO_DRUID_OPERATORS = ImmutableMap.of(
+            "-", "SUB",
+            "+", "ADD",
+            "*", "MULT",
+            "/", "DIV");
+    private static final String FROM_UNIXTIME = "from_unixtime";
+
+    private final FunctionMetadataManager functionMetadataManager;
+    private final ConnectorSession session;
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java]: not used

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    @Override
+    public DruidExpression visitCall(
+            CallExpression call,
+            Map<VariableReferenceExpression, DruidQueryGeneratorContext.Selection> context)
+    {
+        Optional<DruidExpression> basicCallHandlingResult = basicCallHandling(call, context);
+        if (basicCallHandlingResult.isPresent()) {
+            return basicCallHandlingResult.get();
+        }
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java]: break a line

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        FunctionMetadata functionMetadata = functionMetadataManager.getFunctionMetadata(call.getFunctionHandle());
+        Optional<OperatorType> operatorTypeOptional = functionMetadata.getOperatorType();
+        if (operatorTypeOptional.isPresent()) {
+            OperatorType operatorType = operatorTypeOptional.get();
+            if (operatorType.isArithmeticOperator()) {
+                return handleArithmeticExpression(call, operatorType, context);
+            }
+            if (operatorType.isComparisonOperator()) {
+                throw new PrestoException(DRUID_PUSHDOWN_UNSUPPORTED_EXPRESSION, "Unsupported operator: " + call + " to pushdown for Druid connector ");
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java]: nit: replace the space with a period a the end.

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            }
+        }
+        return handleFunction(call, context);
+    }
+
+    @Override
+    public DruidExpression visitConstant(
+            ConstantExpression literal,
+            Map<VariableReferenceExpression, DruidQueryGeneratorContext.Selection> context)
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java]: one param per line; same for all other functions in this class.

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            CallExpression function,
+            Map<VariableReferenceExpression, DruidQueryGeneratorContext.Selection> context)
+    {
+        RowExpression timeInputParameter = function.getArguments().get(1);
+        String inputColumn;
+        String inputTimeZone;
+        String inputFormat;
+
+        CallExpression timeConversion = getExpressionAsFunction(timeInputParameter, timeInputParameter);
+        switch (timeConversion.getDisplayName().toLowerCase(ENGLISH)) {
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java]: Use `if`

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            case FROM_UNIXTIME:
+                inputColumn = timeConversion.getArguments().get(0).accept(this, context).getDefinition();
+                inputTimeZone = timeConversion.getArguments().size() > 1 ? getStringFromConstant(timeConversion.getArguments().get(1)) : DateTimeZone.UTC.getID();
+                inputFormat = "seconds";
+                break;
+            default:
+                throw new PrestoException(DRUID_PUSHDOWN_UNSUPPORTED_EXPRESSION, "Unsupported time function: " + timeConversion.getDisplayName() + " to pushdown for Druid connector ");
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java]: same nit: `.. Druid connector "`; there are many typos in this patch

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+            return derived(format("%s(%s, %s)", druidOperator, left.getDefinition(), right.getDefinition()));
+        }
+        throw new PrestoException(DRUID_PUSHDOWN_UNSUPPORTED_EXPRESSION, "Unsupported arithmetic expression: " + expression + " to pushdown for Druid connector ");
+    }
+
+    private DruidExpression handleFunction(
+            CallExpression function,
+            Map<VariableReferenceExpression, DruidQueryGeneratorContext.Selection> context)
+    {
+        switch (function.getDisplayName().toLowerCase(ENGLISH)) {
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java]: if

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidPushdownUtils.java [hunk 8]
```diff
@@ -42,6 +47,106 @@
 {
     private DruidPushdownUtils() {}
 
+    public enum ExpressionType
+    {
+        GROUP_BY,
+        AGGREGATE,
+    }
+
+    public static class GroupByColumnNode
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidPushdownUtils.java]: Having a public enum in a util class doesn't sound good. Can you make `AggregationColumnNode` a top-level class and call it `DruidAggregationColumnNode`? Put the enum and two implementations as the inner class of it.

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 9]
```diff
@@ -220,9 +243,112 @@ public DruidQueryGeneratorContext visitTableScan(TableScanNode node, DruidQueryG
             Map<VariableReferenceExpression, Selection> selections = new LinkedHashMap<>();
             node.getOutputVariables().forEach(outputColumn -> {
                 DruidColumnHandle druidColumn = (DruidColumnHandle) (node.getAssignments().get(outputColumn));
+                checkArgument(druidColumn.getType().equals(DruidColumnHandle.DruidColumnType.REGULAR), "Unexpected druid column handle that is not regular: " + druidColumn);
                 selections.put(outputColumn, new Selection(druidColumn.getColumnName(), TABLE_COLUMN));
             });
             return new DruidQueryGeneratorContext(selections, tableHandle.getTableName());
         }
+
+        private String handleAggregationFunction(CallExpression aggregation, Map<VariableReferenceExpression, Selection> inputSelections)
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: Move to the end of the class

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 10]
```diff
@@ -220,9 +243,112 @@ public DruidQueryGeneratorContext visitTableScan(TableScanNode node, DruidQueryG
...
+        {
+            String prestoAggregation = aggregation.getDisplayName().toLowerCase(ENGLISH);
+            List<RowExpression> parameters = aggregation.getArguments();
+            switch (prestoAggregation) {
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: if

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 11]
```diff
@@ -220,9 +243,112 @@ public DruidQueryGeneratorContext visitTableScan(TableScanNode node, DruidQueryG
...
+                        return format("count(%s)", parameters.isEmpty() ? "*" : inputSelections.get(getVariableReference(parameters.get(0))));
+                    }
+                    break;
+                default:
+                    if (UNARY_AGGREGATION_MAP.containsKey(prestoAggregation) && aggregation.getArguments().size() == 1) {
+                        return format("%s(%s)", UNARY_AGGREGATION_MAP.get(prestoAggregation), inputSelections.get(getVariableReference(parameters.get(0))));
+                    }
+            }
+
+            throw new PrestoException(DRUID_PUSHDOWN_UNSUPPORTED_EXPRESSION, "Unsupported pushdown for Druid connector. Aggregation function:" + aggregation + " not supported");
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: leave a space after `function:`

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 12]
```diff
@@ -220,9 +243,112 @@ public DruidQueryGeneratorContext visitTableScan(TableScanNode node, DruidQueryG
...
+        }
+
+        @Override
+        public DruidQueryGeneratorContext visitAggregation(AggregationNode node, DruidQueryGeneratorContext contextIn)
+        {
+            List<AggregationColumnNode> aggregationColumnNodes = computeAggregationNodes(node);
+
+            // Make two passes over the aggregatinColumnNodes: In the first pass identify all the variables that will be used
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: typo aggregatin

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java (hunk 4 above)
Comment (zhenxiao) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidAggregationProjectConverter.java]: yep

**Comment by zhenxiao** [comment]
thank you, @highker 
comments addressed

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 13]
```diff
@@ -220,9 +244,108 @@ public DruidQueryGeneratorContext visitTableScan(TableScanNode node, DruidQueryG
...
+                    }
+                    default:
+                        throw new PrestoException(DRUID_PUSHDOWN_UNSUPPORTED_EXPRESSION, "Unsupported pushdown for Druid connector. Unknown aggregation expression:" + expression.getExpressionType());
+                }
+            }
+
+            // Handling non-aggregated group by
+            if (groupByExists && aggregations == 0) {
+                VariableReferenceExpression hidden = new VariableReferenceExpression(UUID.randomUUID().toString(), BigintType.BIGINT);
+                newSelections.put(hidden, new Selection("count(*)", DERIVED));
```
Comment (luohao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: Is `count` the only aggregation function in this case?

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 14]
```diff
@@ -19,33 +19,56 @@
...
 import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+import static java.util.Locale.ENGLISH;
 import static java.util.Objects.requireNonNull;
 
 public class DruidQueryGenerator
 {
     private static final Logger log = Logger.get(DruidQueryGenerator.class);
+    private static final Map<String, String> UNARY_AGGREGATION_MAP = ImmutableMap.of(
```
[the remark this answers is not in the document]
Comment (zhenxiao) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: yep, will add in following PRs. Start with easy ones :)

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 15]
```diff
@@ -220,9 +244,108 @@ public DruidQueryGeneratorContext visitTableScan(TableScanNode node, DruidQueryG
...
+                        newSelections.put(getVariableReference(aggregationNode.getOutputColumn()), new Selection(druidAggregationFunction, DERIVED));
+                        aggregations++;
+                        break;
+                    }
+                    default:
+                        throw new PrestoException(DRUID_PUSHDOWN_UNSUPPORTED_EXPRESSION, "Unsupported pushdown for Druid connector. Unknown aggregation expression:" + expression.getExpressionType());
+                }
+            }
+
+            // Handling non-aggregated group by
```
[the remark this answers is not in the document]
Comment (zhenxiao) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: query with Group By, but no aggregations, e.g. count(*)

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java (hunk 13 above)
Comment (zhenxiao) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: yep, count(*) is the only case

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 16]
```diff
@@ -220,9 +244,108 @@ public DruidQueryGeneratorContext visitTableScan(TableScanNode node, DruidQueryG
...
+                    case GROUP_BY: {
+                        GroupByColumnNode groupByColumn = (GroupByColumnNode) expression;
+                        VariableReferenceExpression groupByInputColumn = getVariableReference(groupByColumn.getInputColumn());
+                        variablesInAggregation.add(groupByInputColumn);
+                        break;
+                    }
+                    case AGGREGATE: {
+                        AggregationFunctionColumnNode aggregationNode = (AggregationFunctionColumnNode) expression;
+                        variablesInAggregation.addAll(
+                                aggregationNode
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: break after `.stream()`
```java
aggregationNode.getCallExpression().getArguments().stream()
        .filter(...
        .map..

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 17]
```diff
@@ -220,9 +244,108 @@ public DruidQueryGeneratorContext visitTableScan(TableScanNode node, DruidQueryG
...
+
+        private String handleAggregationFunction(CallExpression aggregation, Map<VariableReferenceExpression, Selection> inputSelections)
+        {
+            String prestoAggregation = aggregation.getDisplayName().toLowerCase(ENGLISH);
+            List<RowExpression> parameters = aggregation.getArguments();
+            if (prestoAggregation.equals("count")) {
+                if (parameters.size() <= 1) {
+                    return format("count(%s)", parameters.isEmpty() ? "*" : inputSelections.get(getVariableReference(parameters.get(0))));
+                }
+                else if (UNARY_AGGREGATION_MAP.containsKey(prestoAggregation) && aggregation.getArguments().size() == 1) {
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: this is always false

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGeneratorContext.java [hunk 18]
```diff
@@ -140,17 +227,25 @@ private boolean hasFilter()
...
-        if (limit.isPresent()) {
-            query += " LIMIT " + limit.getAsLong();
+        if (!groupByColumns.isEmpty()) {
+            String groupByExpression = groupByColumns.stream().map(x -> selections.get(x).getDefinition()).collect(Collectors.joining(", "));
+            query = query + " GROUP BY " + groupByExpression;
             pushdown = true;
         }
+
+        if (!hasAggregation()) {
+            if (limit.isPresent()) {
```
Comment (highker) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGeneratorContext.java]: merge this two ifs

File: presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java [hunk 19]
```diff
@@ -92,4 +93,31 @@ public void testSimpleSelectWithFilterLimit()
...
     }
+
+    @Test
+    public void testCountStar()
+    {
+        BiConsumer<PlanBuilder, PlanBuilder.AggregationBuilder> aggregationFunctionBuilder = (planBuilder, aggregationBuilder) -> aggregationBuilder.addAggregation(planBuilder.variable("agg"), getRowExpression("count(*)", defaultSessionHolder));
+        PlanNode justScan = buildPlan(planBuilder -> tableScan(planBuilder, druidTable, regionId, secondsSinceEpoch, city, fare));
+        PlanNode filter = buildPlan(planBuilder -> filter(planBuilder, tableScan(planBuilder, druidTable, regionId, secondsSinceEpoch, city, fare), getRowExpression("fare > 3", defaultSessionHolder)));
+        PlanNode anotherFilter = buildPlan(planBuilder -> filter(planBuilder, tableScan(planBuilder, druidTable, regionId, secondsSinceEpoch, city, fare), getRowExpression("secondssinceepoch between 200 and 300 and regionid >= 40", defaultSessionHolder)));
+        testDQL(planBuilder -> planBuilder.aggregation(aggBuilder -> aggregationFunctionBuilder.accept(planBuilder, aggBuilder.source(justScan).globalGrouping())),
```
Comment (highker) [inline presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java]: one param per line

File: presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java [hunk 20]
```diff
@@ -92,4 +93,31 @@ public void testSimpleSelectWithFilterLimit()
...
+                "SELECT regionId, count(*) FROM realtimeOnly GROUP BY regionId");
+        testDQL(planBuilder -> planBuilder.aggregation(aggBuilder -> aggregationFunctionBuilder.accept(planBuilder, aggBuilder.source(anotherFilter).singleGroupingSet(v("regionid"), v("city")))),
+                "SELECT regionId, city, count(*) FROM realtimeOnly WHERE ((secondsSinceEpoch BETWEEN 200 AND 300) AND (regionId >= 40)) GROUP BY regionId, city");
+    }
+
+    @Test
+    public void testDistinctSelection()
+    {
+        PlanNode justScan = buildPlan(planBuilder -> tableScan(planBuilder, druidTable, regionId, secondsSinceEpoch, city, fare));
+        testDQL(planBuilder -> planBuilder.aggregation(aggBuilder -> aggBuilder.source(justScan).singleGroupingSet(v("regionid"))),
```
Comment (highker) [inline presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java]: same

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 21]
```diff
@@ -19,33 +19,56 @@
...
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
 
+import static com.facebook.presto.druid.DruidAggregationColumnNode.AggregationFunctionColumnNode;
+import static com.facebook.presto.druid.DruidAggregationColumnNode.ExpressionType.AGGREGATE;
+import static com.facebook.presto.druid.DruidAggregationColumnNode.ExpressionType.GROUP_BY;
```
[the remark this answers is not in the document]
Comment (zhenxiao) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: ? It is used in a few switch cases

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java (hunk 17 above)
Comment (zhenxiao) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java]: oh, my bad. Let me fix it

File: presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java (hunk 19 above)
Comment (zhenxiao) [reply presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java]: yep, just 2 params for testDQL. current format seems OK?

**Comment by zhenxiao** [comment]
thank you @highker 
comments addressed.

File: presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java (hunk 19 above)
Comment (highker) [reply presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java]: the first line should be empty; the first param starts at the second line. No param needs to follow right after `(`

Comment (zhenxiao) [reply presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java]: get it. will fix

**Comment by zhenxiao** [comment]
@luohao am going to merge this PR soon. Could you please take a look, see if any problems?

**Pull request merged**